app/assets/javascripts/components/samples.jsx

```diff
@@ -434,6 +447,103 @@ class Samples extends React.Component {
     }
   }
 
+  toggleFavorite(e) {
+    let favStatus = e.target.getAttribute('data-fav');
+    let projectId = e.target.getAttribute('data-id');
+    Samples.showLoading(`${favStatus == 'true' ? 'Removing from' : 'Adding to' } favorites...`)
+    axios
+      .put(`/projects/${projectId}/favorite?`, {
+        authenticity_token: this.csrf
+      })
+      .then((res) => {
+        Samples.hideLoader();
+        this.checkIfProjecExistInFavorites(projectId, this.state.formattedProjectList);
+      }).catch((err) => {
+        Samples.hideLoader();
+    })
+  }
+
+  reformatProjectList(favorites, allProjects) {
+    let favProjects = [];
+    let favIds = favorites.map(e => e.id);
+    let formattedList = allProjects.map(e => {
+      let project = e;
+
+      if (!project.favorited) {
+        project.favorited = favIds.includes(project.id);
+
+        if (project.favorited) {
+          favProjects.push(project);
+        }
+      } 
+
+      return project;
+    });
+
+    this.setState({ 
+      formattedProjectList: formattedList,
+      formattedFavProjectList: favProjects,
+      favIds: favIds
+    });
+
+    return formattedList;
+  }
+
+  // update
+  updateProjectsState(id, projects) {
+    let updatedList = projects.map(project => {
+      if (project.id == id) {
+        project.favorited = !project.favorited
+      }
+
+      return project;
+    });
+
+    this.setState({ formattedProjectList: updatedList });
+
+    return updatedList;
+  }
+
+  // check existence of projects in favorites projects list
+  // if true then remove else add
+  checkIfProjecExistInFavorites(id, projects) {
+    if (this.state.favIds.includes(parseInt(id))) {
+      this.removeProjectFromFavorites(id);
+    } else {
+      this.addProjectToFavorites(id, projects);
+    }
+    this.updateProjectsState(id, projects);
+  }
+
+
+  // remove Projects from favorites list
+  removeProjectFromFavorites(id) {
+    let updatedFavouriteProjects = this.state.formattedFavProjectList.filter(project => project.id != id);
+    let removedFavouriteProject = this.state.formattedFavProjectList.filter(project => project.id == id);
+
+    let favIds = this.state.favIds;
+    let projectIdIndex = favIds.indexOf(removedFavouriteProject[0].id);
+
+    if (projectIdIndex > -1) {
+      favIds.splice(projectIdIndex, 1);
+      this.setState({ 
+        formattedFavProjectList: updatedFavouriteProjects,
+          favIds
+      });
+    }
+  }
+
+
+  // add Projects to favorites list
+  addProjectToFavorites(id, projects) {
+    let updatedProject = projects.filter(project => project.id == id);
+
+    this.setState({
+        formattedFavProjectList: [...this.state.formattedFavProjectList, ...updatedProject],
```

Comment:
do you need the second ... here?

Reply:
yes, the second is also an array, we join them using the spread syntax.